From 678a8ce5e0626e235bdf1af346a018ef0d459a98 Mon Sep 17 00:00:00 2001
From: NLObP <38990417+NL0bP@users.noreply.github.com>
Date: Wed, 7 Aug 2024 17:13:00 +0300
Subject: [PATCH] [Game][Packets] Added the SCAccountAttribute2Packet packet:
 In addition, I transferred the sending of the SCScheduledEventStartedPacket
 packet from CSSelectCharacterPacket to CSNotifyInGameCompletedPacket, so that
 the event message is visible on the screen (for example, double exp, etc.).
 Added sending a SCChatMessagePacket with a welcome message for logging into
 Archeage.

---
 .../C2G/CSNotifyInGameCompletedPacket.cs      |  5 +++
 .../Packets/C2G/CSSelectCharacterPacket.cs    |  4 +-
 .../Packets/G2C/SCAccountAttribute2Packet.cs  | 37 +++++++++++++++++++
 AAEmu.Game/Core/Packets/G2C/SCOffsets.cs      |  2 +-
 .../Characters/CharacterOnlineTrackingTask.cs |  3 ++
 5 files changed, 47 insertions(+), 4 deletions(-)
 create mode 100644 AAEmu.Game/Core/Packets/G2C/SCAccountAttribute2Packet.cs

diff --git a/AAEmu.Game/Core/Packets/C2G/CSNotifyInGameCompletedPacket.cs b/AAEmu.Game/Core/Packets/C2G/CSNotifyInGameCompletedPacket.cs
index c5a152cb96..6f913009f0 100644
--- a/AAEmu.Game/Core/Packets/C2G/CSNotifyInGameCompletedPacket.cs
+++ b/AAEmu.Game/Core/Packets/C2G/CSNotifyInGameCompletedPacket.cs
@@ -1,6 +1,9 @@
 ﻿using AAEmu.Commons.Network;
 using AAEmu.Game.Core.Managers.World;
 using AAEmu.Game.Core.Network.Game;
+using AAEmu.Game.Core.Packets.G2C;
+using AAEmu.Game.Models;
+using AAEmu.Game.Models.Game.Chat;
 
 namespace AAEmu.Game.Core.Packets.C2G;
 
@@ -12,6 +15,8 @@ public CSNotifyInGameCompletedPacket() : base(CSOffsets.CSNotifyInGameCompletedP
 
     public override void Read(PacketStream stream)
     {
+        Connection.SendPacket(new SCScheduledEventStartedPacket());
+        Connection.SendPacket(new SCChatMessagePacket(ChatType.System, AppConfiguration.Instance.World.MOTD)); // "Welcome to AAEmu!"
 
         WorldManager.Instance.OnPlayerJoin(Connection.ActiveChar);
         Logger.Info("NotifyInGameCompleted SubZoneId {0}", Connection.ActiveChar.SubZoneId);
diff --git a/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs b/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
index afc5d24638..2cd30a040f 100644
--- a/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
+++ b/AAEmu.Game/Core/Packets/C2G/CSSelectCharacterPacket.cs
@@ -8,6 +8,7 @@
 using AAEmu.Game.Core.Managers.World;
 using AAEmu.Game.Core.Network.Game;
 using AAEmu.Game.Core.Packets.G2C;
+using AAEmu.Game.Models.Game;
 using AAEmu.Game.Models.Game.Char;
 using AAEmu.Game.Models.Game.Items.Actions;
 using AAEmu.Game.Models.Game.Skills;
@@ -116,9 +117,6 @@ public override void Read(PacketStream stream)
 
             character.Breath = character.LungCapacity;
 
-            Connection.SendPacket(new SCScheduledEventStartedPacket());
-            Connection.SendPacket(new SCItemTaskSuccessPacket(ItemTaskType.Unk136, [], []));
-
             Connection.ActiveChar.OnZoneChange(0, Connection.ActiveChar.Transform.ZoneId);
         }
         else
diff --git a/AAEmu.Game/Core/Packets/G2C/SCAccountAttribute2Packet.cs b/AAEmu.Game/Core/Packets/G2C/SCAccountAttribute2Packet.cs
new file mode 100644
index 0000000000..5ad25e3be7
--- /dev/null
+++ b/AAEmu.Game/Core/Packets/G2C/SCAccountAttribute2Packet.cs
@@ -0,0 +1,37 @@
+﻿using System;
+
+using AAEmu.Commons.Network;
+using AAEmu.Game.Core.Network.Game;
+
+namespace AAEmu.Game.Core.Packets.G2C;
+
+public class SCAccountAttribute2Packet : GamePacket
+{
+    private readonly byte _AccountAttributeKind;
+    private readonly uint _extraKind;
+    private readonly byte _worldId;
+    private readonly uint _count;
+    private readonly DateTime _startDate;
+    private readonly DateTime _endData;
+
+    public SCAccountAttribute2Packet() : base(SCOffsets.SCAccountAttribute2Packet, 5)
+    {
+        _AccountAttributeKind = 1;
+        _extraKind = 0;
+        _worldId = 0x1;
+        _count = 0;
+        _startDate = DateTime.Now;
+        _endData = DateTime.MinValue;
+    }
+
+    public override PacketStream Write(PacketStream stream)
+    {
+        stream.Write(_AccountAttributeKind);
+        stream.Write(_extraKind);
+        stream.Write(_worldId);
+        stream.Write(_count);
+        stream.Write(_startDate);
+        stream.Write(_endData);
+        return stream;
+    }
+}
diff --git a/AAEmu.Game/Core/Packets/G2C/SCOffsets.cs b/AAEmu.Game/Core/Packets/G2C/SCOffsets.cs
index 31b587be63..3359a1e6e9 100644
--- a/AAEmu.Game/Core/Packets/G2C/SCOffsets.cs
+++ b/AAEmu.Game/Core/Packets/G2C/SCOffsets.cs
@@ -669,7 +669,7 @@ public static class SCOffsets
     public const ushort SCUnitAiAggroPacket = 0x027;
     public const ushort SCCharacterGenderAndModelModifiedPacket = 0x172;
     public const ushort SCAccountAttributePacket = 0x28C; // off_39C19E24
-    public const ushort off_39C19E30 = 0x112;
+    public const ushort SCAccountAttribute2Packet = 0x112; // off_39C19E30
     public const ushort off_39C19E48 = 0x008;
     public const ushort off_39C19E54 = 0x108;
     public const ushort SCItemLinkTextForLootPackItemBroadcastPacket = 0x2BB;
diff --git a/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs b/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
index e7fbe58667..9fd6b36a96 100644
--- a/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
+++ b/AAEmu.Game/Models/Tasks/Characters/CharacterOnlineTrackingTask.cs
@@ -3,6 +3,7 @@
 using AAEmu.Game.Core.Managers.World;
 using AAEmu.Game.Core.Packets.G2C;
 using AAEmu.Game.Models.Game;
+using AAEmu.Game.Models.Game.Items.Actions;
 
 namespace AAEmu.Game.Models.Tasks.Characters;
 
@@ -58,6 +59,8 @@ public override void Execute()
                     Updated = DateTime.UtcNow
                 };
                 character.SendPacket(new SCScheduleItemUpdatePacket([si]));
+                character.SendPacket(new SCItemTaskSuccessPacket(ItemTaskType.Unk136, [], []));
+
             }
         }
 
